### PR TITLE
Finish cmake configuration to support `FindPackage` and friends

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

Add the package configuration.

## PR Checklist

- [x] Code is formatted.

### CODE REVIEW

**Comment by Yurlungur** [comment]
This now works, thanks to the help of @mauneyc-LANL , and is ready for review.

**Comment by mauneyc-LANL** [comment]
@Yurlungur up to 211e8c9 I redid some directory setup and export config, and updated the test. Let me know if any issues arise on your end

**Comment by mauneyc-LANL** [comment]
> @Yurlungur up to [211e8c9](https://github.com/lanl/ports-of-call/commit/211e8c9164881efcdc0a98384a47e997ee223b80) I redid some directory setup and export config, and updated the test. Let me know if any issues arise on your end

And some tidying up, test now attempts to compile skeleton code that just includes a PoC header.

File: CMakeLists.txt [hunk 1]
```diff
@@ -40,8 +58,32 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
...
+    not valid, defaulting to \"None\", valid options are 
+    ${PORTABILITY_STRATEGY_OPTIONS}")
+  list(GET PORTABILITY_STRATEGY_OPTIONS 0 PORTABILITY_STRATEGY)
+endif()
+# string-manip for the define
+string(TOUPPER ${PORTABILITY_STRATEGY} PORTABILITY_STRATEGY)
+set(PORTABILITY_STRATEGY "PORTABILITY_STRATEGY_${PORTABILITY_STRATEGY}")
+message(STATUS "Will define ${PORTABILITY_STRATEGY}")
+
+mark_as_advanced(PORTABILITY_STRATEGY)
```
Comment (Yurlungur) [inline CMakeLists.txt:82]: What does this do?

**Review by Yurlungur — COMMENTED** [review]
I like this. Thanks @mauneyc-LANL !

**Comment by Yurlungur** [comment]
Nice improvements, @mauneyc-LANL ! Thanks!

One thought: This implies that the portability strategy is set at "install" time. I guess this is consistent with how, e.g., Kokkos works. And it doesn't do anything regarding in-tree builds. Is being able to change the portability strategy "post-install" something that's important to people? I think since in-tree builds are still flexible, this is consistent with, e.g., Kokkos, and probably fine. 

What are people's thoughts?

**Comment by Yurlungur** [comment]
@chadmeyer @mauneyc-LANL @junghans @jdolence @brryan I'd like to merge this soon. I'll wait 24 hours and if no one has reviewed by then, I will merge by fiat.

File: .github/workflows/install.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        - name: install dependencies
+          run: |
+            sudo apt-get install -y --force-yes -qq build-essential
+        - name: build and install
+          run: |
+            mkdir -p build
+            cd build
+            mkdir -p ${HOME}/install
+            echo "Build"
+            cmake -DCMAKE_INSTALL_PREFIX=${HOME}/install ..
```
Comment (junghans) [inline .github/workflows/install.yml:30]: Remove `mkdir` and `cd` above and just use `cmake -B build` here and `cmake --build build` instead of `make`

**Review by junghans — CHANGES_REQUESTED** [review]

**Review by chadmeyer — APPROVED** [review]
I think this looks reasonable to me.  Thanks to everyone who has pulled this together!

File: .github/workflows/install.yml (hunk 2 above)
Comment (Yurlungur) [reply .github/workflows/install.yml]: Good point. Done.

**Comment by Yurlungur** [comment]
@junghans I've addressed your comment. Please take a look (and if you're happy) approve when you get the chance.

**Comment by mauneyc-LANL** [comment]
> Nice improvements, @mauneyc-LANL ! Thanks!
> 
> One thought: This implies that the portability strategy is set at "install" time. I guess this is consistent with how, e.g., Kokkos works. And it doesn't do anything regarding in-tree builds. Is being able to change the portability strategy "post-install" something that's important to people? I think since in-tree builds are still flexible, this is consistent with, e.g., Kokkos, and probably fine.
> 
> What are people's thoughts?

If this is for downstream configuration, it probably shouldn't propagate the usage requirement at all. The code in `portability` sorts out all the logic on it's own wherever it's included from.

File: CMakeLists.txt [hunk 3]
```diff
@@ -39,38 +58,102 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
...
-  target_compile_definitions(ports-of-call::ports-of-call
-    INTERFACE
-    PORTABILITY_STRATEGY_NONE)
-  message(STATUS "No portability strategy set. Setting to None")
+# CONFIGURATION LOGIC
+# ----------------------------------------
+
+# TODO: I ended up not liking this
+# set the portability strategy. we only allow 3 options: None, Kokkos, Cuda
+set(PORTABILITY_STRATEGY_OPTIONS "None;Kokkos;Cuda")
```
Comment (junghans) [inline CMakeLists.txt:66]: Add:
```
set_property(CACHE PORTABILITY_STRATEGY PROPERTY STRINGS ${PORTABILITY_STRATEGY_OPTIONS})
```
so it shows up in `ccache`.

File: CMakeLists.txt [hunk 4]
```diff
@@ -39,38 +58,102 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
...
+if(NOT PORTABILITY_STRATEGY)
+  list(GET PORTABILITY_STRATEGY_OPTIONS 0 PORTABILITY_STRATEGY)
 endif()
+# if the cache variable is not valid, display a warning and default to none
+list(FIND PORTABILITY_STRATEGY_OPTIONS ${PORTABILITY_STRATEGY} _psidx)
+if(_psidx EQUAL -1)
```
Comment (junghans) [inline CMakeLists.txt:72]: Modern cmake is:
```
if(PORTABILITY_STRATEGY IN_LIST PORTABILITY_STRATEGY_OPTIONS)
```

File: .github/workflows/install.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+          run: |
+            echo "Build"
+            mkdir -p ${HOME}/install
+            cmake -DCMAKE_INSTALL_PREFIX=${HOME}/install -B build .
+            cmake --build build
+            cmake --install build
+            echo "Now try find package from a separate directory"
+            rm -rf build
+            cd installtest
+            cmake -DCMAKE_PREFIX_PATH=${HOME}/install -B build .
```
Comment (junghans) [inline .github/workflows/install.yml:34]:
```suggestion
            cmake -DCMAKE_PREFIX_PATH=${HOME}/install -B build installtest
```
and drop `cd` above

File: ports-of-callConfig.cmake.in [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# - Config file for the PortsOfCall package
+# It defines the following variables
+#  FOOBAR_INCLUDE_DIRS - include directories for FooBar
+#  FOOBAR_LIBRARIES    - libraries to link against
+
+@PACKAGE_INIT@
+
+include("${CMAKE_CURRENT_LIST_DIR}/ports-of-callTargets.cmake")
```
Comment (junghans) [inline ports-of-callConfig.cmake.in:8]:
```suggestion
include("${CMAKE_CURRENT_LIST_DIR}/@POCLIB@Targets.cmake")
```

File: ports-of-callConfig.cmake.in [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+check_required_components(ports-of-call)
```
Comment (junghans) [inline ports-of-callConfig.cmake.in:10]:
```suggestion
check_required_components(@POCLIB@)
```

**Review by junghans — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 3 above)
Comment (Yurlungur) [reply CMakeLists.txt:66]: :+1:

File: CMakeLists.txt (hunk 4 above)
Comment (Yurlungur) [reply CMakeLists.txt:72]: :+1:

File: .github/workflows/install.yml (hunk 5 above)
Comment (Yurlungur) [reply .github/workflows/install.yml:34]: :+1:

**Review by junghans — APPROVED** [review]

**Comment by Yurlungur** [comment]
All change requests addressed. Also passed the `PORTABILITY_STRATEGY` through spack as a variant and added some guards against multiple portability strategies. Merging. Thanks all for your help!

**Pull request merged**